### PR TITLE
Personal/issue05

### PR DESCRIPTION
**Pull request by Debskij**
Finished game board on player side and on viewers side

### CODE REVIEW

File: client/game_logic.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        self.package_number = 0
+        self.score = 0
+        self.combo = 1
+        self.number_of_moles = 9
+        self.values_changed = False
+        self.mole_life_time = [2, 3]
+        self.max_moles_amount = 3
+        self.last_print_time = 0
+        self.moles_status = [[0, [0, 0, 0]]
+                             for _ in range(self.number_of_moles)]  # [status, alive_time]
```
Comment (krzysztofzajaczkowski) [inline client/game_logic.py:24]: What is alive_time? Isn't it just a number dictating when a mole hides?
How many states moles have?

File: client/game_logic.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+                    if self.score >= MAX_SCORE:
+                        self.end = True
+                    print(self.create_bits_package())
+                time.sleep(0.2)
+        except KeyboardInterrupt:
+            return
+
+    def generate_mole_lifetime(self) -> list:
+        value = random.uniform(self.mole_life_time[0], self.mole_life_time[1])
+        return [time.time() + value * (x + 1) / 3 for x in range(3)]
```
Comment (krzysztofzajaczkowski) [inline client/game_logic.py:49]: Why there are 3 numbers representing alive_time?

File: client/game_logic.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+    def set_new_mole(self, pos: int):
+        self.moles_status[pos] = [3, self.generate_mole_lifetime()]
```
Comment (krzysztofzajaczkowski) [inline client/game_logic.py:52]: What does status 3 mean?

File: client/game_logic.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+    def moles_progression(self):
+        self.moles_lock.acquire()
+        op_start_time = time.time()
+        for idx, mole in enumerate(self.moles_status):
+            for state, checked_time in enumerate(mole[1][::-1]):
+                if checked_time < op_start_time:
+                    self.moles_status[idx][0] = state
+                    self.values_changed = True
+                    break
```
Comment (krzysztofzajaczkowski) [inline client/game_logic.py:65]: Shouldn't alive_time be altered/erased when moles state has changed?

File: client/game_logic.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        self.pressed_key = []
+        self.pressed_keys_lock.release()
+        self.moles_lock.acquire()
+        for key in unique_keys:
+            self.values_changed = True
+            self.calculate_score(key)
+        self.moles_lock.release()
+
+    def calculate_score(self, key: int, max_combo=8):
+        if sum(self.moles_status[key]) != 0:
```
Comment (krzysztofzajaczkowski) [inline client/game_logic.py:92]: What does sum(self.moles_status[key]) dictates? Don't we need just its status to determine if player hit the target successfully?
Furthermore, wouldn't it sum up alive_time too?

File: client/game_logic.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            self.combo = 1
+
+    def print_board(self):
+        self.moles_lock.acquire()
+        row_length = int(math.sqrt(self.number_of_moles))
+        for idx in range(row_length):
+            print("\t".join([str(x[0])
+                             for x in self.moles_status[idx * row_length:(idx + 1) * row_length]]))
+        print("\n")
+        self.moles_lock.release()
```
Comment (krzysztofzajaczkowski) [inline client/game_logic.py:106]: It's just for debugging purpose for now, right?

**Review by krzysztofzajaczkowski — CHANGES_REQUESTED** [review]
I'd like You to tell more about moles_status, since it's a little bit confusing.

File: client/game_logic.py (hunk 1 above)
Comment (Debskij) [reply client/game_logic.py:24]: moles hold 4 status actually, 3 is straight when they appear, after 1/3 of live time goes out it should change to 2, after 2/3 to 1 after whole time to 0.

File: client/game_logic.py (hunk 2 above)
Comment (Debskij) [reply client/game_logic.py:49]: Up

File: client/game_logic.py (hunk 3 above)
Comment (Debskij) [reply client/game_logic.py:52]: Up

File: client/game_logic.py (hunk 4 above)
Comment (Debskij) [reply client/game_logic.py:65]: Nope, thats why it has 3 time gates. Each one means 1/3 of life time. Thats why i check it in reversed order

File: client/game_logic.py (hunk 5 above)
Comment (Debskij) [reply client/game_logic.py:92]: After mole dies, its life time goes to [0, 0, 0] besides that sum(int, list) = sum(int) if im correct

File: client/game_logic.py (hunk 6 above)
Comment (Debskij) [reply client/game_logic.py:106]: Yes, i'd suggest to remove it as soon as you will be totally sure everything here works fine

**Review by krzysztofzajaczkowski — APPROVED** [review]

**Comment by Debskij** [comment]
As requested, moles_status:
FORMAT:
For each mole existing in game we hold list containing int and list of ints [status_of_mole, [first_level_hide, second_level_hide, third_level_hide]] or simply [status_of_mole, time_array]
VARIABLES:
status_of_mole (type: int) may only represent one of 4 states (0, 1, 2, 3). When mole appears on board its status is set by default on 3. It means it should be fully revealed. Status 1 and 2 are reserved for moles hiding after time. 0 is special status when mole is totally hidden. Whenever status_of_mole is equal to 0 sum of time_array should also be equal to 0. 
time_array (type: list(time, time, time)) its set when mole appears. each time is formed by adding current time to formula (idx+1)/3*mole_life_time. When third part of mole life time passes, which means that current time is "bigger" than time_array[0] it should change mole status from 3 to 2 etc.
KEEP IN MIND:
moles_status are divided between multiple functions, every time you want to use this variable you should remember to use moles_lock (type: threading.Lock) to reduce hazard unexpected problems. 

In case of any further problems feel free to hit me up

**Pull request merged**